packages/toolpad-app/src/toolpad/DomLoader.tsx

```diff
@@ -75,8 +82,51 @@ export type DomAction =
       node: appDom.AppDomNode;
     };
 
+const undoStack: appDom.AppDom[] = [];
+const redoStack: appDom.AppDom[] = [];
+
+const UNDO_LIMIT = 100;
+
+const updateUndoStack = throttle((dom: appDom.AppDom) => {
+  undoStack.push(dom);
+  // Destroy redo stack when new action is executed
+  redoStack.length = 0;
+
+  if (undoStack.length > UNDO_LIMIT) {
+    undoStack.shift();
+  }
+}, 500);
+
+const SKIP_UNDO_ACTIONS = ['DOM_UNDO', 'DOM_REDO', 'DOM_SAVED', 'DOM_SAVING', 'DOM_SAVING_ERROR'];
```

Comment:
I'd always use a `Set` if an array is used for lookup. Remember that `[].includes` has O(n) complexity and `new Set().has` has O(1) complexity. 

The impact is probably pretty small in this case, but it's good form to use the right tool for the job.

Reply:
Thanks, updated this 👍